src/lib/storage/table.cpp

```diff
@@ -373,39 +403,114 @@ void Table::create_chunk_index(const std::vector<ColumnID>& column_ids, const st
   _chunk_indexes_statistics.emplace_back(indexes_statistics);
 }
 
-const TableKeyConstraints& Table::soft_key_constraints() const {
-  return _table_key_constraints;
-}
-
 void Table::add_soft_key_constraint(const TableKeyConstraint& table_key_constraint) {
-  Assert(_type == TableType::Data, "Key constraints are not tracked for reference tables across the PQP.");
+  Assert(_type == TableType::Data, "TableKeyConstraints are not tracked for reference tables across the PQP.");
 
-  // Check validity of specified columns
+  // Check validity of specified columns.
+  const auto column_count = this->column_count();
   for (const auto& column_id : table_key_constraint.columns()) {
-    Assert(column_id < column_count(), "ColumnID out of range");
+    Assert(column_id < column_count, "ColumnID out of range.");
 
-    // PRIMARY KEY requires non-nullable columns
+    // PRIMARY KEY requires non-nullable columns.
     if (table_key_constraint.key_type() == KeyConstraintType::PRIMARY_KEY) {
       Assert(!column_is_nullable(column_id), "Column must be non-nullable to comply with PRIMARY KEY.");
     }
   }
 
-  {
-    auto scoped_lock = acquire_append_mutex();
+  const auto append_lock = acquire_append_mutex();
 
-    for (const auto& existing_constraint : _table_key_constraints) {
-      // Ensure that no other PRIMARY KEY is defined
-      Assert(existing_constraint.key_type() == KeyConstraintType::UNIQUE ||
-                 table_key_constraint.key_type() == KeyConstraintType::UNIQUE,
-             "Another primary key already exists for this table.");
+  for (const auto& existing_constraint : _table_key_constraints) {
+    // Ensure that no other PRIMARY KEY is defined.
+    Assert(existing_constraint.key_type() == KeyConstraintType::UNIQUE ||
+               table_key_constraint.key_type() == KeyConstraintType::UNIQUE,
+           "Another primary TableKeyConstraint exists for this table.");
 
-      // Ensure there is only one key constraint per column set.
-      Assert(table_key_constraint.columns() != existing_constraint.columns(),
-             "Another key constraint for the same column set has already been defined.");
```

Comment:
This would accept Unique constraints like `[0]` and [0,1]`, right? Don't we want to enforce "maximal constraints"?

Reply:
In most cases, we want minimal constraints (if a column is unique, also the projection on this column plus another columns is unique). Anyways, this is only a whitespace change. So you're somehow right, we should for now assume the columns do not intersect.